Apps/phredmine/redmine_connector.py

```diff
@@ -0,0 +1,996 @@
+            file_contents = f.read()
+
+        ret_val, response = self._make_rest_call(
+            "/uploads.json",
+            action_result,
+            params=None,
+            method="post",
+            headers={"Content-Type": "application/octet-stream"},
+            data=file_contents,
+        )
+
+        if phantom.is_fail(ret_val):
+            return action_result.set_status(
+                phantom.APP_ERROR, REDMINE_ERR_UPLOAD_ATTACHMENT
+            )
+
+        upload = {
+            "token": response["upload"]["token"],
+            "filename": file_info["name"],
+            "content_type": file_info["mime_type"],
+        }
+
+        return upload
+
+    # Action Handlers
+
+    def _handle_test_connectivity(self, param):
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        self.save_progress("Connecting to endpoint")
+        ret_val, _ = self._make_rest_call(
+            "/my/account.json", action_result, params=None, headers=None
+        )
+
+        if phantom.is_fail(ret_val):
+            self.debug_print(action_result.get_message())
+            self.save_progress(REDMINE_ERR_TEST_CONN_FAILED)
+            return action_result.set_status(phantom.APP_ERROR)
+
+        self.save_progress(REDMINE_SUCC_TEST_CONN_PASSED)
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def _on_poll(self, param):
+        """
+        Keep phantom containers up-to-date with data from redmine
+        """
+
+        # Add action result
+        action_result = self.add_action_result(phantom.ActionResult(dict(param)))
+
+        last_run = self._state.get("last_run")
+        max_tickets = None
+        backfill = datetime.now() - timedelta(REDMINE_BACKFILL_DAYS)
+
+        if self.is_poll_now():
+            self.debug_print("Run Mode: Poll Now")
+
+            # Integer validation for 'maximum containers' configuration parameter
+            max_tickets = param[phantom.APP_JSON_CONTAINER_COUNT]
+            ret_val, max_tickets = self._validate_integer(action_result, max_tickets, REDMINE_CONTAINER_COUNT_KEY)
+            if phantom.is_fail(ret_val):
+                return action_result.get_status()
+            last_run = backfill
+        else:
+            if not last_run:
+                self.debug_print("Run Mode: First Scheduled Poll")
+                last_run = backfill
+            else:
+                self.debug_print("Run Mode: Scheduled Poll")
+                last_run = dateutil.parser.isoparse(last_run)
+
+        self.debug_print(f"Last Run: {last_run}")
+
+        tickets = []
+        try:
+            ret_val, tickets = self._retrieve_tickets(
+                action_result, last_run, max_tickets
+            )
+            if phantom.is_fail(ret_val):
+                return action_result.get_status()
+        except Exception as e:
+            error_msg = self._get_error_message_from_exception(e)
+            action_result.set_status(
```

Comment:
Can action be returned from here, or should we leave the code lines as they are? And if we don't have a return declaration, the action will be marked as successful with 0 total tickets later in the code.

Reply:
@dfederschmidt, can you help with this?

Reply:
@arathore-crest yes, we can return action_result in this case.

```suggestion
            return action_result.set_status(phantom.APP_ERROR, REDMINE_ERR_FETCHING_TICKETS.format(error=error_msg))
```

Reply:
Added return statement